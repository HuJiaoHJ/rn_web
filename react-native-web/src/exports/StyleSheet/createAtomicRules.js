import createKeyframesRules from './createKeyframesRules';
import createRuleBlock from './createRuleBlock';

const createAtomicRules = (selector, prop, value) => {
  const rules = [];

  // Handle custom properties and custom values that require additional rules
  // to be created.
  switch (prop) {
    // See #513
    case 'pointerEvents': {
      let val = value;
      if (value === 'auto' || value === 'box-only') {
        val = 'auto !important';
        if (value === 'box-only') {
          const block = createRuleBlock({ [prop]: 'none' });
          rules.push(`${selector} > *{${block}}`);
        }
      } else if (value === 'none' || value === 'box-none') {
        val = 'none !important';
        if (value === 'box-none') {
          const block = createRuleBlock({ [prop]: 'auto' });
          rules.push(`${selector} > *{${block}}`);
        }
      }
      const block = createRuleBlock({ [prop]: val });
      rules.push(`${selector}{${block}}`);
      break;
    }

    case 'placeholderTextColor': {
      const block = createRuleBlock({ color: value, opacity: 1 });
      rules.push(
        '@media all {' +
          `${selector}::-webkit-input-placeholder{${block}}` +
          `${selector}::-moz-placeholder{${block}}` +
          `${selector}:-ms-input-placeholder{${block}}` +
          `${selector}::placeholder{${block}}` +
          '}'
      );
      break;
    }

    case 'animationName': {
      if (typeof value === 'string') {
        // add a className referencing the animation
        const block = createRuleBlock({ [prop]: value });
        rules.push(`${selector}{${block}}`);
      } else {
        const animationNames = [];

        // add the keyframes needed to implement each value
        value.forEach(keyframes => {
          if (typeof keyframes === 'string') {
            animationNames.push(keyframes);
          } else {
            const { identifier, rules: keyframesRules } = createKeyframesRules(keyframes);
            keyframesRules.forEach(rule => {
              rules.push(rule);
            });
            animationNames.push(identifier);
          }
        });

        // add a className referencing the animation identifiers
        const block = createRuleBlock({ [prop]: animationNames.join(',') });
        rules.push(`${selector}{${block}}`);
      }

      break;
    }

    default: {
      const block = createRuleBlock({ [prop]: value });
      rules.push(`${selector}{${block}}`);
    }
  }

  return rules;
};

export default createAtomicRules;
