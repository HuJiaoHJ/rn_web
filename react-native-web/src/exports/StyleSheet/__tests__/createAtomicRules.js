/* eslint-env jasmine, jest */

import createAtomicRules from '../createAtomicRules';

describe('StyleSheet/createAtomicRules', () => {
  test('transforms standard declarations to a single rule', () => {
    expect(createAtomicRules('.test', 'margin', 0)).toMatchSnapshot();
  });

  test('transforms custom "animationName" declaration', () => {
    const value = [
      { '0%': { top: 0 }, '50%': { top: 5 }, '100%': { top: 10 } },
      { from: { left: 0 }, to: { left: 10 } }
    ];
    expect(createAtomicRules('.test', 'animationName', value)).toMatchSnapshot();
  });

  test('transforms custom "pointerEvents" declaration', () => {
    expect(createAtomicRules('.test', 'pointerEvents', 'box-only')).toMatchSnapshot();
  });

  test('transforms custom "placeholderTextColor" declaration', () => {
    expect(createAtomicRules('.test', 'placeholderTextColor', 'gray')).toMatchSnapshot();
  });
});
