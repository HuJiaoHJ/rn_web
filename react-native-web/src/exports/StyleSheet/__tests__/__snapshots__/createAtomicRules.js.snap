// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StyleSheet/createAtomicRules transforms custom "animationName" declaration 1`] = `
Array [
  "@media all {@-webkit-keyframes rn-anim-2k74q5{0%{top:0px}50%{top:5px}100%{top:10px}}}",
  "@media all {@keyframes rn-anim-2k74q5{0%{top:0px}50%{top:5px}100%{top:10px}}}",
  "@media all {@-webkit-keyframes rn-anim-zc91cv{from{left:0px}to{left:10px}}}",
  "@media all {@keyframes rn-anim-zc91cv{from{left:0px}to{left:10px}}}",
  ".test{-webkit-animation-name:rn-anim-2k74q5,rn-anim-zc91cv;animation-name:rn-anim-2k74q5,rn-anim-zc91cv}",
]
`;

exports[`StyleSheet/createAtomicRules transforms custom "placeholderTextColor" declaration 1`] = `
Array [
  "@media all {.test::-webkit-input-placeholder{color:gray;opacity:1}.test::-moz-placeholder{color:gray;opacity:1}.test:-ms-input-placeholder{color:gray;opacity:1}.test::placeholder{color:gray;opacity:1}}",
]
`;

exports[`StyleSheet/createAtomicRules transforms custom "pointerEvents" declaration 1`] = `
Array [
  ".test > *{pointer-events:none}",
  ".test{pointer-events:auto !important}",
]
`;

exports[`StyleSheet/createAtomicRules transforms standard declarations to a single rule 1`] = `
Array [
  ".test{margin:0px}",
]
`;
